health/health.go

```diff
@@ -476,31 +499,60 @@ func (t *Tracker) AppendWarnableDebugFlags(base []string) []string {
 	return ret
 }
 
-// RegisterWatcher adds a function that will be called whenever the health state of any Warnable changes.
-// If a Warnable becomes unhealthy or its unhealthy state is updated, the callback will be called with its
-// current Representation.
-// If a Warnable becomes healthy, the callback will be called with ws set to nil.
-// The provided callback function will be executed in its own goroutine. The returned function can be used
-// to unregister the callback.
-func (t *Tracker) RegisterWatcher(cb func(w *Warnable, r *UnhealthyState)) (unregister func()) {
-	return t.registerSyncWatcher(func(w *Warnable, r *UnhealthyState) {
-		go cb(w, r)
+// Change is used to communicate a change to health. This could either be due to
+// a Warnable changing from health to unhealthy (or vice-versa), or because the
+// health messages received from the control-plane have changed.
+type Change struct {
+	// Health messages from the POV of the control-plane server changed. If set, WarnableChanged, Warnable, and UnhealthyState are not set.
+	ControlHealthChanged bool
+
+	// A client Warnable changed state. Mutually exclusive with ControlHealthChanged. If set, Warnable is set and ControlHealthChanged is not set.
+	WarnableChanged bool
+
+	// Warnable that changed state. If it is now unhealthy, UnhealthyState will
+	// be set.
+	Warnable *Warnable
+
+	// UnhealthyState is set if the Warnable is unhealthy, and unset if the
+	// Warnable is now healthy.
+	UnhealthyState *UnhealthyState
+}
+
+// RegisterWatcher adds a function that will be called whenever the health state
+// of any Warnable changes or the health messages from the control-plane change.
+//
+// If a Warnable becomes unhealthy or its unhealthy state is updated, the
+// callback will be called with WarnableChanged set to true and the Warnable and
+// its UnhealthyState.
+//
+// If a Warnable becomes healthy, the callback will be called with
+// WarnableChanged set to true, the Warnable set, and UnhealthyState set to nil
+//
+// If the health messages from the control-plane change, the callback will be
+// called with ControlHealthChanged set to true. Clients can fetch the set of
+// control-plane health messages by calling [Tracker.CurrentState].
+//
+// The provided callback function will be executed in its own goroutine. The
+// returned function can be used to unregister the callback.
+func (t *Tracker) RegisterWatcher(cb func(c Change)) (unregister func()) {
+	return t.registerSyncWatcher(func(c Change) {
+		go cb(c)
```

Comment:
Firing off goroutines always seems to bite us via flaky tests or ordering problems or resource leaks on shutdown.

(That's why we made goroutines.Tracker and execqueue)

But I guess this is what the code did earlier, so it's fine.